htmd/qm/psi4.py

```diff
@@ -0,0 +1,244 @@
+    >>> from htmd.queues.acecloudqueue import AceCloudQueue
+    >>> qm.queue = AceCloudQueue() # doctest: +SKIP
+    """
+
+    @property
+    def _command(self):
+        return 'export HTMD_PSI4_WORKDIR=$(pwd)\npsi4 -i psi4.in -o psi4.out &> psi4.log'
+
+    def _completed(self, directory):
+        return os.path.exists(os.path.join(directory, 'psi4.out'))
+
+    def _writeInput(self, directory, iframe):
+
+        with open(os.path.join(directory, 'psi4.in'), 'w') as f:
+
+            f.write('set_num_threads( %d )\n' % self._ncpus)
+            f.write('memory %f gb\n\n' % self._mem)
+
+            reference = 'r' if self.multiplicity == 1 else 'u'
+            reference += 'hf' if self.theory == 'RHF' else 'ks'
+            f.write('set { reference %s }\n' % reference)
+            f.write('set { basis %s }\n\n' % self.basis)
+
+            if self.solvent == 'vacuum':
+                pass
+            elif self.solvent == 'PCM':
+                # TODO check all agruments
+                f.write('set { PCM true\n      PCM_scf_type total }\n')
+                f.write('PCM = {\n')
+                f.write('  Units = Angstrom\n')
+                f.write('  Medium {\n    SolverType = IEFPCM\n    Solvent = Water\n  }\n')
+                f.write('  Cavity {\n    RadiiSet = UFF\n    Type = GePol\n')
+                f.write('    Scaling = False\n    Area = 0.3\n    Mode = Implicit\n  }\n')
+                f.write('}\n\n')
+            else:
+                raise NotImplementedError
+
+            # Write the molecule
+            f.write('molecule MOL {\n    %d %d\n' % (self._charge, self.multiplicity))
+            elements = self._molecule.element
+            coords = self._molecule.coords[:, :, iframe]
+            for element, coord in zip(elements, coords):
+                f.write('    %-2s %10f %10f %10f\n' % (element, coord[0], coord[1], coord[2]))
+            f.write('    symmetry c1 }\n\n')
+
+            if self._restrained_dihedrals is not None:
+                dihedrals = ['%d %d %d %d' % tuple(dihedral) for dihedral in self._restrained_dihedrals]
+                dihedrals = ', '.join(dihedrals)
+                f.write('set optking { frozen_dihedral = (" %s ") }\n\n' % dihedrals)
+
+            # Enable a dynamic optimization algorithm selection to converge problematic cases:
+            # http://www.psicode.org/psi4manual/master/optking.html#dealing-with-problematic-optimizations
+            if self.optimize:
+                f.write('set optking { dynamic_level = 1 }\n\n')
+
+            theory_correction = 'SCF' if self.theory == 'HF' else self.theory
+            theory_correction += '' if self.correction == 'none' else '-%s' % self.correction
+            function = 'optimize' if self.optimize else 'energy'
+            f.write('energy, wfn = %s(\'%s\', return_wfn=True)\n\n' % (function, theory_correction))
+
+            # Psi4 changes directory then PCM is used, so we need to return
```

Comment:
God, this is ugly... Isn't there another way of doing this?

Reply:
This is kind of a bug in Psi4!

Reply:
Maybe we could ping them on this and keep this behaviour for now